R/tmb.R

```diff
@@ -41,31 +41,32 @@ h_mmrm_tmb_control <- function(optimizer = stats::nlminb,
   )
 }
 
-#' Processing the covariance term
-#'
-#' @param cov_term (`call`)\cr covariance term.
-#'
-#' @return Named list of with elements:
+#' Extract group/subject vars from covariance term
+#' @param cov_content (`call`)\cr covariance term
+#' @details a covariance term, which takes the following form:
+#' `us(time | group / subject)`, or `sp_exp(time1, time2, ... | group / subject)`.
+#' Only for spatial covariance structure can have multiple time coordinates.
+#' This function, take the call of `time | group / subject`, last element of the covariance term,
+#' extract the corresponding grouping var, time var and subject var.
+#' @return Named list with elements:
 #' - `visit_var`: `string` with the visit variable name.
 #' - `subject_var`: `string` with the subject variable name.
 #' - `group_var`: `string` with the group variable name. If no group specified, this element
 #'      is NULL.
 #' @keywords internal
-h_mmrm_tmb_extract_vars <- function(cov_term) {
-  assert_true(identical(length(cov_term), 2L)) # 2 because `fun (...)`.
-  cov_content <- cov_term[[2L]]
+h_mmrm_tmb_extract_terms <- function(cov_content) {
   if (!identical(length(cov_content), 3L) || !identical(as.character(cov_content[[1L]]), "|")) {
     stop(
-      "Covariance structure must be of the form `",
-      cov_term[[1]],
-      "(time|(group/)subject)`"
+      "Covariance structure must be of the form `time|(group/)subject`."
     )
   }
   visit_term <- cov_content[[2L]]
   if (!identical(length(visit_term), 1L)) {
     stop(
-      "`time` in `(time|(group/)subject)` must be specified as one single variable."
+      "`time` in `time|(group/)subject` must be specified as one single variable."
```

Comment:
for spatial cov structure, e.g.,  `sp_exp(t1, t2, t3 | group / subject)`, the last part of this call, `t3 | group / subject` is used in this function. the `visit_term` is `t3`. The last part is consistent between spatial covariance structure and non-spatial. The `t3` should be a simple symbol instead of a composite like `(t3+t4)`.  So here I think it is fine as `time` indicate the visit term.

forgive my English vocabulary maybe I did not express myself clearly.